File: matlab_impl/+topology/Node.m
classdef Node < handle
    properties
        Id (1,1) uint32
        RefMicId (1,1) uint32   % 参考麦克风 ID
        SecSpkId (1,1) uint32   % 次级扬声器 ID
        ErrMicId (1,1) uint32   % 误差麦克风 ID
        NeighborIds (1,:) uint32     % 存储邻居节点 ID

    end

    methods
        function obj = Node(id)
            if nargin > 0
                obj.Id = id;
                obj.NeighborIds = uint32(id); 
            end
        end

        function addRefMic(obj, micId)
            obj.RefMicId = micId;
        end

        function addSecSpk(obj, spkId)
            obj.SecSpkId = spkId;
        end

        function addErrMic(obj, micId)
            obj.ErrMicId = micId;
        end
    end
end

File: py_anc/topology/node.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List


@dataclass
class Node:
    """Physical node mapping one reference mic, one secondary speaker and one error mic."""

    node_id: int
    ref_mic_id: int | None = None
    sec_spk_id: int | None = None
    err_mic_id: int | None = None
    neighbor_ids: List[int] = field(default_factory=list)

    def __post_init__(self) -> None:
        if not self.neighbor_ids:
            self.neighbor_ids = [self.node_id]

    def add_ref_mic(self, mic_id: int) -> None:
        self.ref_mic_id = mic_id

    def add_sec_spk(self, spk_id: int) -> None:
        self.sec_spk_id = spk_id

    def add_err_mic(self, mic_id: int) -> None:
        self.err_mic_id = mic_id
